File: src/main/java/com/company/HanoiPanel.java
package com.company;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.Arrays;

public class HanoiPanel extends JPanel {
    private static final Color POLE_COLOR = new Color(0x594700);
    private static final Font STEP_FONT = new Font("Monospace", Font.PLAIN, 15);
    private static final char[] STEP_TEXT = new char[9];
    static {
        // initialize the first few chars
        STEP_TEXT[0] = 'S';
        STEP_TEXT[1] = 't';
        STEP_TEXT[2] = 'e';
        STEP_TEXT[3] = 'p';
        STEP_TEXT[4] = ' ';
    }
    private ArrayList<Disk> stackA;
    private final ArrayList<Disk> stackB = new ArrayList<>();
    private final ArrayList<Disk> stackC = new ArrayList<>();
    private final Disk[] disks;
    private final TowerOfHanoi solver;

    private int curr;

    public HanoiPanel(int n) {
        super();
        setBackground(Color.white);
        stackA = new ArrayList<>(n);
        disks = new Disk[n];
        for (int i = 0; i < n; i++) {
            disks[i] = new Disk(n - i - 1);
        }
        solver = new TowerOfHanoi(n);
        reset();
    }

    public boolean end() {
        return curr == solver.getSize() - 1;
    }

    public void moveTo(int num) {
        if (num == 1) {
            reset();
        } else {
            while (curr + 2 != num) {
                if (curr + 2 > num) {
                    prev();
                } else {
                    next();
                }
            }
        }
        repaint();
    }
    private void next() {
        if (curr == solver.getSize() - 1) return;
        TowerOfHanoi.Step step = solver.getStep(++curr);
        Disk curr = switch (step.from) {
            case 'A' -> stackA.remove(stackA.size() - 1);
            case 'B' -> stackB.remove(stackB.size() - 1);
            case 'C' -> stackC.remove(stackC.size() - 1);
            default -> throw new IllegalStateException("Unexpected value: " + step.from);
        };
        switch (step.to) {
            case 'A' -> stackA.add(curr);
            case 'B' -> stackB.add(curr);
            case 'C' -> stackC.add(curr);
        }
    }
    private void prev() {
        if (curr < 0) return;
        TowerOfHanoi.Step step = solver.getStep(curr--);
        Disk curr = switch (step.to) {
            case 'A' -> stackA.remove(stackA.size() - 1);
            case 'B' -> stackB.remove(stackB.size() - 1);
            case 'C' -> stackC.remove(stackC.size() - 1);
            default -> throw new IllegalStateException("Unexpected value: " + step.from);
        };
        switch (step.from) {
            case 'A' -> stackA.add(curr);
            case 'B' -> stackB.add(curr);
            case 'C' -> stackC.add(curr);
        }
    }
    public void reset() {
        curr = -1;
        stackA = new ArrayList<>(Arrays.asList(disks));
        stackB.clear();
        stackC.clear();
        repaint();
    }

    @Override
    public void paintComponent(Graphics g) {
        super.paintComponent(g);
        int width = getWidth();
        int height = getHeight();
        // draw poles
        g.setColor(POLE_COLOR);
        for (int i = 0, j = width / 4; i < 3; i++, j += width / 4) {
            g.fillRoundRect(j, height - Disk.HEIGHT * 11, 5, Disk.HEIGHT * 12, 5, 5);
        }
        // draw disks
        for (int i = 0; i < stackA.size(); i++) {
            stackA.get(i).draw(g, i, 0, width, height);
        }
        for (int i = 0; i < stackB.size(); i++) {
            stackB.get(i).draw(g, i, 1, width, height);
        }
        for (int i = 0; i < stackC.size(); i++) {
            stackC.get(i).draw(g, i, 2, width, height);
        }
        // draw step
        g.setColor(Color.black);
        String number = Integer.toString(curr + 1);
        for (int i = 0; i < number.length(); i++) {
            STEP_TEXT[i + 5] = number.charAt(i);
        }
        g.setFont(STEP_FONT);
        g.drawChars(STEP_TEXT, 0, number.length() + 5, 5, 20);
    }

    private static class Disk {
        public static final int HEIGHT = 20;
        public static final int INITIAL_WIDTH = 50;
        private final int number;
        private static final Color[] COLORS = {
                new Color(0xff8e8e),
                new Color(0xecb283),
                new Color(0xd8ce77),
                new Color(0xa7c56c),
                new Color(0x72b261),
                new Color(0x569f68),
                new Color(0x4b8c78),
                new Color(0x417179),
                new Color(0x364a66),
                new Color(0x2d2c52)
        };
        private final Color color;
        private static final int WIDTH_INC = 10;
        public Disk(int number) {
            this.number = number;
            color = COLORS[number];
        }

        /**
         * Draws this rectangle
         * @param g      graphics object
         * @param num    height from bottom starting from 0
         * @param stack  number from 0 to 2, the stack of this rectangle
         * @param width  width of outer panel
         * @param height height of outer panel
         */
        public void draw(Graphics g, int num, int stack, int width, int height) {
            int x = width / 4 * (stack + 1) - (number * WIDTH_INC) - INITIAL_WIDTH / 2;
            int rectWidth = number * WIDTH_INC * 2 + INITIAL_WIDTH;
            g.setColor(color);
            //noinspection SuspiciousNameCombination
            g.fillRoundRect(x, height - num * HEIGHT - HEIGHT, rectWidth, HEIGHT, HEIGHT, HEIGHT);
        }
    }
}

File: src/main/java/com/company/HanoiPanel.kt
package com.company

import javax.swing.JPanel
import java.awt.*

/**
 * Underlying representation of hanoi configuration using 3 stacks
 * @param n number of disks
 */
class HanoiPanel(n: Int) : JPanel() {
    /**
     * Number of disks
     */
    private val nDisks: Int
    private var stackA: ArrayList<Int>
    private val stackB = ArrayList<Int>()
    private val stackC = ArrayList<Int>()
    private val solver: TowerOfHanoi
    private var curr = 0

    init {
        background = Color.white
        stackA = ArrayList(n)
        nDisks = n
        solver = TowerOfHanoi(n)
        reset()
    }

    /**
     * Whether there are no more steps
     * @return true if there are no more steps
     */
    fun end(): Boolean {
        return curr == solver.size - 1
    }

    /**
     * Change the panel to show the nth step
     * @param num the step to show, 1 being the initial configuration
     */
    fun moveTo(num: Int) {
        if (num == 1) {
            reset()
        } else {
            while (curr + 2 != num) {
                if (curr + 2 > num) {
                    prev()
                } else {
                    next()
                }
            }
        }
        repaint()
    }

    /**
     * Go to next step
     */
    private operator fun next() {
        if (curr == solver.size - 1) return
        val step = solver.getStep(++curr)!!
        val curr = when (step.from) {
            'A' -> stackA.removeAt(stackA.size - 1)
            'B' -> stackB.removeAt(stackB.size - 1)
            'C' -> stackC.removeAt(stackC.size - 1)
            else -> throw IllegalStateException("Unexpected value: " + step.from)
        }
        when (step.to) {
            'A' -> stackA.add(curr)
            'B' -> stackB.add(curr)
            'C' -> stackC.add(curr)
            else -> throw IllegalStateException("Unexpected value: " + step.to)
        }
    }

    /**
     * Go to previous step
     */
    private fun prev() {
        if (curr < 0) return
        val step = solver.getStep(curr--)!!
        val curr = when (step.to) {
            'A' -> stackA.removeAt(stackA.size - 1)
            'B' -> stackB.removeAt(stackB.size - 1)
            'C' -> stackC.removeAt(stackC.size - 1)
            else -> throw IllegalStateException("Unexpected value: " + step.to)
        }
        when (step.from) {
            'A' -> stackA.add(curr)
            'B' -> stackB.add(curr)
            'C' -> stackC.add(curr)
            else -> throw IllegalStateException("Unexpected value: " + step.from)
        }
    }

    /**
     * Resets the disks to the starting position
     */
    fun reset() {
        curr = -1
        // fill stackA with decreasing numbers
        stackA = ArrayList(generateSequence(nDisks - 1){it - 1}.take(nDisks).toList())
        stackB.clear()
        stackC.clear()
        repaint()
    }

    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        // turn on antialiasing
        (g as Graphics2D).setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
        val width = width
        val height = height
        // draw poles
        g.setColor(POLE_COLOR)
        run {
            var i = 0
            var j = width / 4 - 2
            while (i < 3) {
                g.fillRoundRect(j, height - D_HEIGHT * (MAX + 1), 5, D_HEIGHT * (MAX + 2), 5, 5)
                i++
                j += width / 4
            }
        }
        // draw disks
        for (i in stackA.indices) {
            drawDisk(g, i, 0, width, height, stackA[i])
        }
        for (i in stackB.indices) {
            drawDisk(g, i, 1, width, height, stackB[i])
        }
        for (i in stackC.indices) {
            drawDisk(g, i, 2, width, height, stackC[i])
        }
        // draw step
        g.setColor(Color.black)
        g.setFont(STEP_FONT)
        g.drawString("Step " + (curr + 1), 5, 20)
    }

    /**
     * Draws a disk
     * @param g      graphics object
     * @param num    height from bottom starting from 0
     * @param stack  number from 0 to 2, the stack this disk is on
     * @param width  width of outer panel
     * @param height height of outer panel
     * @param diskNo the disk number, starting with 0 for the smallest
     */
    private fun drawDisk(g: Graphics, num: Int, stack: Int, width: Int, height: Int, diskNo: Int) {
        val x = width / 4 * (stack + 1) - diskNo * D_WIDTH_INC - INITIAL_WIDTH / 2
        val rectWidth = diskNo * D_WIDTH_INC * 2 + INITIAL_WIDTH
        g.color = D_COLORS[diskNo]
        g.fillRoundRect(x, height - num * D_HEIGHT - D_HEIGHT, rectWidth, D_HEIGHT, D_HEIGHT, D_HEIGHT)
    }

    companion object {
        /**
         * Max number of disks, so poles are tall enough
         */
        private const val MAX = 10

        /**
         * Font for step text
         */
        private val STEP_FONT = Font("Monospace", Font.PLAIN, 15)

        /**
         * Height of disk
         */
        const val D_HEIGHT = 20

        /**
         * Width of smallest disk
         */
        const val INITIAL_WIDTH = 50

        /**
         * How much the width increases with each successive disk
         */
        private const val D_WIDTH_INC = 10

        /**
         * Color of poles
         */
        private val POLE_COLOR = Color(0x594700)

        /**
         * Color of disks
         */
        private val D_COLORS = arrayOf(
                Color(0xff8e8e),
                Color(0xecb283),
                Color(0xd8ce77),
                Color(0xa7c56c),
                Color(0x72b261),
                Color(0x569f68),
                Color(0x4b8c78),
                Color(0x417179),
                Color(0x364a66),
                Color(0x2d2c52)
        )
    }
}